### ISSUE TITLE
Good strategy for a frontend

### ISSUE THREAD

**Issue by bigfatbird** [jaytaph/gosub-browser#42]
I was wondering what the initial architecture for the frontend/ui could be like? 

Should you create a native hi for each major os, or write once run everywhere like electron (you said you didn't want that)?

I could have a look into a simple macOS gui in the next few days, so there could be something clickable

**Comment by jaytaph** [comment]
I think there are a few things we need to take into account:

  - we (will) have a gosub-engine that does the most work
  - we have user-agents that will display the data and act with user input

I haven't looked into how engines like webkit and gecko and even chromium do this, and how they communicate with the user agents. My initial idea was to see if we can create a (relatively simple) text user agent, as this sounds the least work, but a graphical user-interface is ultimately what it should become.  

I sounds like a browser should be native, to give the most flexibility and performance, but I think we should define first what would be part of the engine, and what would be part of the user agent, and how we would communicate between them.

**Comment by Jesusdasilva2001** [comment]
I think a strategy is a plan of actions designed to achieve a long term or overall aim

**Comment by emwalker** [comment]
I'd pitch for investigating these topics before getting too far into this question:
- [ ] How does Chromium interact with the app layer?
- [ ] How does Firefox interact with the app layer?
- [ ] How does Servo interact with the app layer?

It seems like an easy thing to paint yourself into a corner with if one's assumptions are wrong.

**Comment by jaytaph** [comment]
Agreed with @emwalker. A lot more research is needed on many different topics, but I wouldn't be afraid to experiment with different components at this moment. Writing small proof of concepts gives us just like doing research some insights in the problems we will face later on. 

I don't expect a complete rewrite of the codebase in say a year, but I do think that with knowledge gathered during the year, we should rewrite some parts in order to give it a more stable foundation.

I don't want to create a million pages, but it might be nice if we have a place where we can bundle research (links, notes etc) together. I'm not sure that github is the correct place for this though.
